### PR TITLE
[WIP] Alternative for JoinWith

### PR DESCRIPTION
**Pull request by hvanhovell**

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala [hunk 1]
```diff
@@ -207,12 +207,18 @@ object UnresolvedTVFAliases {
 /**
  * Holds the name of an attribute that has yet to be resolved.
  */
-case class UnresolvedAttribute(nameParts: Seq[String]) extends Attribute with Unevaluable {
+case class UnresolvedAttribute(
+    nameParts: Seq[String],
+    optionalExprId: Option[ExprId])
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala:212]: why do we need this?

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -830,6 +830,66 @@ class Dataset[T] private[sql] (
...
+   * @param other Right side of the join.
+   * @param condition Join expression.
+   * @param joinType Type of join to perform. Default `inner`. Must be one of:
+   *                 `inner`, `cross`, `outer`, `full`, `fullouter`,`full_outer`, `left`,
+   *                 `leftouter`, `left_outer`, `right`, `rightouter`, `right_outer`.
+   *
+   * @group typedrel
+   * @since 3.5.0
+   */
+  def joinWith[U](other: Dataset[U], condition: Column, joinType: String): Dataset[(T, U)] = {
```
Comment (cloud-fan) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:854]: shall we add a new logical plan `UnresolvedJoinWith` to simplify the client implementation?

Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:854]: That is the thing. That would mean I also have to add this to the protocol. It is okay I guess. However, we already have tried that in a different PR (not the new logical plan), and my main gripe with this is that we are adding a very very scala specific thing to both the protocol and the analyzer, for something that can easily be done on the client with a number of primitives (access by ordinal & explicit references). I could probably live without the latter, if we make ordinal access a bit more flexible (I need access to the first and last column of the df).

Comment (cloud-fan) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala:854]: Then can we add that expression to the protocol to do positional column lookup? I'm a bit worried about making the column resolution more complex with `UnresolvedAttribute`. We already have special expressions `GetColumnByOrdinal` and `GetViewColumnByNameAndOrdinal` to find-tune the column lookup logic and we can add more.

**Comment by hvanhovell** [comment]
We are going to go with a protocol change.

**Pull request closed without merging**